### PR TITLE
`Webd` and `elektrad` bulk kdb meta-set

### PR DESCRIPTION
**Pull request by tmakar**
## Basics

- [ ] Short descriptions of your changes are in the release notes
      (added as entry in `doc/news/_preparation_next_release.md` which contains `_(my name)_`)
      **Please always add them to the release notes.**
- [ ] Details of what you changed are in commit messages
      (first line should have `module: short statement` syntax)
- [ ] References to issues, e.g. `close #X`, are in the commit messages.
- [ ] The buildservers are happy. If not, fix **in this order**:
  - add a line in `doc/news/_preparation_next_release.md`
  - reformat the code with `scripts/dev/reformat-all`
  - make all unit tests pass
  - fix all memleaks
  - fix the CI itself (or rebase if already fixed)
- [ ] The PR is rebased with current master.

## Checklist

- [ ] I added unit tests for my code
- [ ] I fully described what my PR does in the documentation
      (not in the PR description)
- [ ] I fixed all affected documentation (see [Documentation Guidelines](https://www.libelektra.org/devgettingstarted/documentation))
- [ ] I fixed all affected decisions (see [Decision Process](https://www.libelektra.org/decisions/decision-process))
- [ ] I added code comments, logging, and assertions as appropriate (see [Coding Guidelines](https://www.libelektra.org/devgettingstarted/coding))
- [ ] I updated all meta data (e.g. README.md of plugins and [METADATA.ini](https://master.libelektra.org/doc/METADATA.ini))
- [ ] I mentioned [every code](/.reuse/dep5) not directly written by me in [reuse syntax](https://reuse.software/)

## Review

- [ ] Documentation is introductory, concise, good to read and describes everything what the PR does
- [ ] Examples are well chosen and understandable
- [ ] Code is conforming to [our Coding Guidelines](https://master.libelektra.org/doc/CODING.md)
- [ ] APIs are conforming to [our Design Guidelines](https://master.libelektra.org/doc/DESIGN.md)
- [ ] Code is consistent to [our Design Decisions](https://master.libelektra.org/doc/decisions)

## Labels

- [ ] Add the "work in progress" label if you do not want the PR to be reviewed yet.
- [ ] Add the "ready to merge" label **if everything is done** and no further pushes are planned by you.

### CODE REVIEW

**Comment by tmakar** [comment]
This PR introduces:

- Implement bulk meta key creation request (the new request supports to add multiple meta keys to  a key in one request)
- Add shell recorder test in README for this new implementation
- Fix existing shell-recorder test to use correct key name

> `NOTE:` Implementation is done for `elektrad` (GO) and for `webd` (NodeJS) .

File: src/tools/elektrad/meta_handler.go [hunk 1]
```diff
@@ -96,16 +102,105 @@ func (s *server) postMetaHandler(w http.ResponseWriter, r *http.Request) {
...
+//
+//	keyName the name of the key. URL path param.
+//	metaSet the set of metakeys for the given keyName
+//
+// Response Code:
+//
+//	201 No Content if the request is successful.
+//	401 Bad Request if no key name was passed - or the key name is invalid.
+//
+// Example: `curl -X POST -d '{ metaSet: [{"key": "hello", "value": "world"}] }' localhost:33333/kdbMeta/user/test/hello`
```
Comment (atmaxinger) [inline src/tools/elektrad/meta_handler.go:118]:
```suggestion
// Example: `curl -X POST -d '{ metaSet: [{"key": "hello", "value": "world"}] }' localhost:33333/kdbMetaBulk/user/test/hello`
```

File: doc/tutorials/install-webui.md [hunk 2]
```diff
@@ -165,6 +165,44 @@ will now return the value of the specified key `user:/test`, which is stored in
 
 <!-- prettier-ignore-end -->
 
+The command
+
+```sh
+curl -X POST -H "Content-Type: application/json" -d '{"metaSet": [{"key": "metakey1", "value": "value1"},{"key": "metakey2", "value": "value2"}]}' http://localhost:33333/kdbMetaBulk/user:/test
```
Comment (atmaxinger) [inline doc/tutorials/install-webui.md:171]: Why is it called `metaSet` and not just `meta`?

Comment (tmakar) [reply doc/tutorials/install-webui.md:171]: No specific reason :smile:. Will change it to `meta`.

**Review by atmaxinger — APPROVED** [review]
For me this looks okay from a pure idea standpoint. However, I can't review the code, as I've never seen a single line of Go code in my life.

It's also a bit off-putting that there are no tests, but apparently, there are no existing tests for elektrad or webd either.

**Comment by tmakar** [comment]
> It's also a bit off-putting that there are no tests, but apparently, there are no existing tests for elektrad or webd either.

Yes, I noticed this as well. 

I now thought that it is better to put at least a shell-recorder test into the `README` than no test at all.

I will may in the next iterations work on adding tests, at least for the `elektrad` tool.

File: src/tools/elektrad/meta_handler.go [hunk 3]
```diff
@@ -96,16 +102,105 @@ func (s *server) postMetaHandler(w http.ResponseWriter, r *http.Request) {
...
+        err = set(handle, ks, errKey)
+
+        if err != nil {
+        	writeError(w, err)
+        	return
+        }
+	}
+
+	noContent(w)
+}
```
Comment (hannes99) [inline src/tools/elektrad/meta_handler.go:190]:
```
func (s *server) postMetaBulkHandler(w http.ResponseWriter, r *http.Request) {
    metaKeySet, keyName, err := parseAndDecodeRequest(r)
    if err != nil {
        writeError(w, err)
        return
    }

    errKey, parentKey, err := createKeys(keyName)
    if err != nil {
        writeError(w, err)
        return
    }
    defer errKey.Close()
    defer parentKey.Close()

    handle, ks := getHandle(r)

    if err := getKeyFromHandle(handle, ks, errKey); err != nil {
        writeError(w, err)
        return
    }

    k := getKeyFromKeySet(ks, keyName, parentKey)

    if err := updateMetaKeys(handle, ks, errKey, k, metaKeySet.Meta); err != nil {
        writeError(w, err)
        return
    }

    noContent(w)
}
```

I didn't work a lot with `Go`, but I feel like creating some smaller functions could improve readability. Other than that the PR LGTM

**Review by hannes99 — APPROVED** [review]

**Comment by tmakar** [comment]
@atmaxinger There actually are tests for `elektrad` but not for `webd`. I was wrong. `meta_handler_test.go`

**Comment by atmaxinger** [comment]
please rebase to resolve merge conflict

**Comment by atmaxinger** [comment]
jenkins build libelektra please

File: doc/news/_preparation_next_release.md [hunk 4]
```diff
@@ -339,6 +339,8 @@ This section keeps you up-to-date with the multi-language support provided by El
 - Added Hannes Laimer to `AUTHORS.md` _(Hannes Laimer @hannes99)_
 - Add README to tools/kdb _(Hannes Laimer @hannes99)_
 - <<TODO>>
+- Fixed shel-recorder test in [`install-webui.md`](../tutorials/install-webui.md) _(Tomislav Makar @tmakar)_
```
Comment (markus2330) [inline doc/news/_preparation_next_release.md:342]:
```suggestion
- Fixed shell-recorder test in [`install-webui.md`](../tutorials/install-webui.md) _(Tomislav Makar @tmakar)_
```

**Review by markus2330 — APPROVED** [review]
LGTM

**Pull request merged**